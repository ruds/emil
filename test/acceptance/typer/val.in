val foo = "hi there";
val foo as (bar, baz) = (foo, 23);
val foo as (bar, _) = (2.5, 23);
val (bar, 23) = (2.5, 23);
val (bar, 23) = (2.5, 23i);
val {a, b=(foo, bar), c=_} = {a="hi", b=(true, 25i), c=(bar, foo)};
val {a, b} = {a="hi", b="there", c="oops"};
val {a, b, ...} = {a="hi", b="there", c="world"};
val {a, b, ...} = {b="where's a?", c="oops again"};
val [a, b, c] = [1, 2, 3];
val [a, b, c] = [1, 2];
val [a, b, c] = [1, 2, 3, 4];
val (::) (a, b) = [1, 2];
val (::) (a, b) = [1];
val (::) (a, b) = [];
val (::) (a, nil) = [];
val (::) (a, nil) = [1];
val (::) (a, nil) = [1, 2];
val (a, ref) = (1, 2);
val ref a = ref 2;
val ref a = ref [];
let val ref a = ref [] in a end;
(*dummy definitions*)
val (+) = fn (a : bigint, _ : bigint) => a;
val (-) = fn (a : bigint, _ : bigint) => a;
val rec f = fn 0 => 1
             | 1 => 1
             | x => (+) (f ((-) (x, 1)), f ((-) (x, 2)));
val map = fn f =>
    let
        val rec g = fn nil => nil
                     | ((::) (el, rest)) => (::) (f el, g rest)
    in
        g
    end;
val rec foldl = fn f => fn z =>
            fn nil => z
             | ((::) (x, xs)) => foldl f (f (z, x)) xs;
val sum = foldl (+) 0;
val rec even = fn 0 => true | n => odd ((-) (n, 1))
    and odd  = fn 0 => false | n => even ((-) (n, 1));
val 'a id = fn (z: 'a) => z;
val ('a, 'b) pair = fn (x: 'a, y: 'b) => (x, y);
val 'a rec f = fn nil : 'a list => [] : 'a list | ((::) (x, xs)) => f xs
       and foo = nil : 'a list;
val 'a foo = let val 'a bar = baz in bar end;
