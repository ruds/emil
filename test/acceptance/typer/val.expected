@0001:
val foo : string
@0002:
val bar : string
val baz : bigint
val foo : (string * bigint)
@0003:
val bar : float
val foo : (float * bigint)
@0004: WARNING
@0004:
val bar : float
@0005: WARNING
@0005: ERROR
@0006:
val a : string
val bar : int
val foo : bool
@0007: ERROR
@0008:
val a : string
val b : string
@0009: ERROR
@0010: WARNING
@0010:
val a : bigint
val b : bigint
val c : bigint
@0011: WARNING
@0011:
val a : bigint
val b : bigint
val c : bigint
@0012: WARNING
@0012:
val a : bigint
val b : bigint
val c : bigint
@0013: WARNING
@0013:
val a : bigint
val b : bigint list
@0014: WARNING
@0014:
val a : bigint
val b : bigint list
@0015: WARNING
@0015:
val a : 'a
val b : 'a list
@0016: WARNING
@0016:
val a : 'a
@0017: WARNING
@0017:
val a : bigint
@0018: WARNING
@0018:
val a : bigint
@0019: ERROR
@0020:
val a : bigint
@0021: WARNING
@0021:
val a : X1 list
@0022:
